lib/spack/llnl/util/lock.py

```diff
@@ -36,6 +37,133 @@
+
+    The ``fcntl`` locks that Spack uses are associated with an inode and a process.
+    This is convenient, because if a process exits, it releases its locks.
+    Unfortunately, this also means that if you close a file, *all* locks associated
+    with that file's inode are released, regardless of whether the process has any
+    other open file descriptors on it.
+
+    Because of this, we need to track open lock files so that we only close them when
+    a process no longer needs them.  We do this by tracking each lockfile by its
+    inode and process id.  This has several nice properties:
+
+    1. Tracking by pid ensures that, if we fork, we don't inadvertently track the parent
+       process's lockfiles. ``fcntl`` locks are not inherited across forks, so we'll
+       just track new lockfiles in the child.
+    2. Tracking by inode ensures that referencs are counted per inode, and that we don't
+       inadvertently close a file whose inode still has open locks.
+    3. Tracking by both pid and inode ensures that we only open lockfiles the minimum
+       number of times necessary for the locks we have.
+
+    Note: as mentioned elsewhere, these locks aren't thread safe -- they're designed to
+    work in Python and assume the GIL.
+    """
+
+    def __init__(self):
+        """Create a new ``OpenFileTracker``."""
+        self._descriptors = {}  # type: Dict[Tuple[int, int], OpenFile]
+
+    def get_fh(self, path, op):
+        """Get a filehandle for a lockfile.
+
+        This routine will open writable files for read/write even if you're asking
+        for a shared (read-only) lock. This is so that we can upgrade to an exclusive
+        (write) lock later if requested.
+
+        Arguments:
+          path (str): path to lock file we want a filehandle for
+          op (int): ``fcntl`` lock operation, e.g., ``fcntl.LOCK_SH``
+        """
+        # Open writable files as 'r+' so we can upgrade to write later
+        os_mode, fh_mode = (os.O_RDWR | os.O_CREAT), 'r+'
+
+        pid = os.getpid()
+        open_file = None  # OpenFile object, if there is one
+        stat = None       # stat result for the lockfile, if it exists
+
+        try:
+            stat = os.stat(path)
+
+            # we know path exists but not if it's writable. If it's read-only,
+            # only open the file for reading (and fail if we're trying to get
+            # an exclusive (write) lock on it)
+            if not os.access(path, os.W_OK):
+                if op == fcntl.LOCK_SH:
+                    # can still lock read-only files if we open 'r'
+                    os_mode, fh_mode = os.O_RDONLY, 'r'
+                else:
+                    # can't do LOCK_EX for read-only files
+                    raise LockROFileError(path)
+
+            # see whether we've seen this inode/pid before
+            key = (stat.st_ino, pid)
+            open_file = self._descriptors.get(key)
```

Comment:
Curious: Why check if file already seen *after* check access and determine modes? Might there be a more meaningful exception if we *have* seen it but no longer have the desired access for example?

Reply:
Well, it doesn't make sense to check again -- or at least there is no need.  If you have a file descriptor open for writing, and the file's permissions change so that you would no longer be able to open it for writing, you can still write to it as long as you hang on to the descriptor.  I suppose we could check again if you're upgrading from read to write to see if we could re-open the file, but I think it's not worth it to do multiple opens in the same spack run.  They can just run spack again.

That said, this brings up another point, which is that we can avoid the call to `os.access()` if the file is already open.  My original intent with this block of code was to use the stat result to check permissions, but it turns out implementing `os.access` is complicated, in that the stat result just tells you whether the owner, group, or other can read/write/etc. -- not whether *you* can.  I think it's better to just use `os.access()`, but it's not necessary to do that until we try to actually open the file.  I'll refactor!

Reply:
Ok @tldahlgren I pushed a change that separates concerns a bit better and avoids the redundancy you noticed above.  I think the logic is easier to follow, and `get_fh` now does not need to know about lock operations -- those are only known to `_lock`, which I think is an improvement.  This version is also ever so slightly faster as it avoids the call to `os.access()` altogether if the file is already open. See what you think now.

@adamjstewart can you check one more time?  This version should be functionally identical but it's worth another check.

Reply:
I'm seeing bad things on this commit:
```console
Error: Failed to install libiconv due to TypeError: cannot pickle '_io.TextIOWrapper' object
```
Not sure if this is from this PR or from #21215. I'm guessing from this PR since I've never seen it before.

P.S. We _really_ need to solve #21215 before we make clingo the default concretizer.

Reply:
Switching back to the old concretizer for now so I can actually test this PR (and use Spack on macOS)

Reply:
I suspect that is #21215. I think we should start making macOS CI just use clingo.  Maybe we should do that first.

Reply:
There's a lot of things we need to start testing on macOS, `spack test` still doesn't work for the same reason.